### PR TITLE
[feat] App Circuit Utils for Keccak Coprocessor

### PR DESCRIPTION
**Pull request by nyunyunyunyu**
- Encoding for `VarLenBytesVec` and `FixLenBytesVec`.

### CODE REVIEW

**Review by jonathanpwang — APPROVED** [review]
Left some minor comments, but LGTM.

BTW just to sanity check: the choice of little endian for keccak byte -> word but big endian for poseidon u64 -> big word was just practical right? I thought about it and it didn't seem there were any issues with leading or trailing zeros because we encode the length.

File: hashes/zkevm/src/keccak/coprocessor/encode.rs [hunk 1]
```diff
@@ -69,7 +79,60 @@ pub fn encode_native_input<F: Field>(bytes: &[u8]) -> F {
...
+pub fn encode_var_len_bytes_vec<F: Field>(
+    ctx: &mut Context<F>,
+    range_chip: &impl RangeInstructions<F>,
+    initialized_hasher: &PoseidonHasher<F, POSEIDON_T, POSEIDON_RATE>,
+    bytes: &VarLenBytesVec<F>,
+) -> AssignedValue<F> {
+    let max_len = bytes.max_len();
+    let max_num_keccak_f = get_num_keccak_f(max_len);
+    // num_keccak_f = len / NUM_BYTES_TO_ABSORB + 1
+    let num_bits = (usize::BITS - max_len.leading_zeros()) as usize;
```
[the remark this answers is not in the document]
Comment (nyunyunyunyu) [reply hashes/zkevm/src/keccak/coprocessor/encode.rs:92]: Do you mean renaming `num_bits` to `bit_length`? `num_bits` comes from the argument name of `div_mod`.

[the remark this answers is not in the document]
Comment (jonathanpwang) [reply hashes/zkevm/src/keccak/coprocessor/encode.rs:92]: no i meant `usize::BITS - max_len.leading_zeros()` is the same as `bit_length(max_len)` where `bit_length` is a halo2-base utils function

[the remark this answers is not in the document]
Comment (nyunyunyunyu) [reply hashes/zkevm/src/keccak/coprocessor/encode.rs]: Oh TIL

**Comment by nyunyunyunyu** [comment]
> Left some minor comments, but LGTM.
> 
> BTW just to sanity check: the choice of little endian for keccak byte -> word but big endian for poseidon u64 -> big word was just practical right? I thought about it and it didn't seem there were any issues with leading or trailing zeros because we encode the length.

@jonathanpwang  Why is u64 -> big word big-endian? let's say bytes is [0x01, 0x02, .., 0x18]. Then it's [0x0807060504030201, 0x0f0e0d0c0b0a09, 0x1817161514131211] -> 0x18171615141312110f0e0d0c0b0a090807060504030201. It's always little endian.

**Comment by jonathanpwang** [comment]
Ah ok, I misread because you had the `fold` iterator.

**Pull request merged**